ulib/libax/src/cbindings/io_mpx/epoll.rs

```diff
@@ -0,0 +1,240 @@
+        Err(LinuxError::ENOSYS)
+    }
+
+    fn set_nonblocking(&self, _nonblocking: bool) -> LinuxResult {
+        Ok(())
+    }
+}
+
+/// `ax_epoll_create()` creates a new epoll instance.
+///
+/// `ax_epoll_create()` returns a file descriptor referring to the new epoll instance.
+#[no_mangle]
+pub unsafe extern "C" fn ax_epoll_create(size: c_int) -> c_int {
+    debug!("ax_epoll_create <= {}", size);
+    ax_call_body!(ax_epoll_create, {
+        if size < 0 {
+            return Err(LinuxError::EINVAL);
+        }
+        let epoll_instance = EpollInstance::new(0);
+        add_file_like(Arc::new(epoll_instance))
+    })
+}
+
+/// Control interface for an epoll file descriptor
+#[no_mangle]
+pub unsafe extern "C" fn ax_epoll_ctl(
+    epfd: c_int,
+    op: c_int,
+    fd: c_int,
+    event: *mut ctypes::epoll_event,
+) -> c_int {
+    debug!("ax_epoll_ctl <= epfd: {} op: {} fd: {}", epfd, op, fd);
+    ax_call_body!(ax_epoll_ctl, {
+        let ep = get_file_like(epfd)?;
+        let ret = ep
+            .into_any()
+            .downcast::<EpollInstance>()
+            .map_err(|_| LinuxError::EINVAL)?
+            .control(op as usize, fd as usize, &(*event))? as c_int;
+        Ok(ret)
+    })
+}
+
+/// `ax_epoll_wait()` waits for events on the epoll instance referred to by the file descriptor epfd.
+#[no_mangle]
+pub unsafe extern "C" fn ax_epoll_wait(
+    epfd: c_int,
+    events: *mut ctypes::epoll_event,
+    maxevents: c_int,
+    timeout: c_int,
+) -> c_int {
+    debug!(
+        "ax_epoll_wait <= epfd: {}, maxevents: {}, timeout: {}",
+        epfd, maxevents, timeout
+    );
+
+    ax_call_body!(ax_epoll_wait, {
+        let epoll_instance = {
+            get_file_like(epfd)?
+                .into_any()
+                .downcast::<EpollInstance>()
+                .map_err(|_| LinuxError::EINVAL)?
+        };
+
+        {
+            epoll_instance.ready_list.lock().clear();
+            epoll_instance
+                .ready_list
+                .lock()
+                .extend(epoll_instance.new_ctl_list.lock().clone());
+            epoll_instance.new_ctl_list.lock().clear();
+        }
+
+        // Add already active events into ready_list
+        for (k, v) in epoll_instance.events.lock().iter() {
+            if (v.events & ctypes::EPOLLET) == 0 {
+                match get_file_like(*k as i32) {
+                    Err(_) => {
+                        return Err(LinuxError::EINVAL);
+                    }
+                    Ok(file_like) => {
+                        let status = file_like.poll()?;
+                        if status.writable || status.readable {
+                            let mut ready_list = epoll_instance.ready_list.lock();
+                            ready_list.insert(*k);
+                        }
+                    }
+                }
+            }
+        }
+
+        let deadline = Some(current_time() + Duration::from_micros(timeout as u64));
```

Comment:
Check if `timeout < 0`.

Reply:
Since when `timeout < 0 ` it is block definitely, move this check into :
```
if timeout >= 0 && deadline.map_or(false, |ddl| current_time() >= ddl) {
                debug!("    timeout!");
                return Ok(0);
            }
```